File: v1/fsharp/minesweeper/MinesweeperTest.fs
module MinesweeperTest

open NUnit.Framework

open Minesweeper

let concat = List.reduce (fun x y -> x + "\n" + y)

[<Test>]
let ``Zero size board`` () =
    let actual = ""
    let expected = ""
    Assert.That(annotate actual, Is.EqualTo(expected))

[<Test>]
let ``Empty board`` () =
    let actual = 
        ["   ";
         "   ";
         "   "]
        |> concat

    let expected = 
        ["   ";
         "   ";
         "   "]
        |> concat

    Assert.That(annotate actual, Is.EqualTo(expected))

[<Test>]
let ``Board full of mines`` () =
    let actual = 
        ["***";
         "***";
         "***"]
        |> concat

    let expected = 
        ["***";
         "***";
         "***"]
        |> concat

    Assert.That(annotate actual, Is.EqualTo(expected))

[<Test>]
let ``Surrounded`` () =    
    let actual = 
        ["***";
         "* *";
         "***"]
        |> concat

    let expected = 
        ["***";
         "*8*";
         "***"]
        |> concat

    Assert.That(annotate actual, Is.EqualTo(expected))

[<Test>]
let ``Horizontal line`` () =    
    let actual = [" * * "] |> concat
    let expected = ["1*2*1"] |> concat

    Assert.That(annotate actual, Is.EqualTo(expected))

[<Test>]
let ``Vertical line`` () =   
    let actual = 
        [" ";
         "*";
         " ";
         "*";
         " "]
        |> concat

    let expected = 
        ["1";
         "*";
         "2";
         "*";
         "1"]
        |> concat

    Assert.That(annotate actual, Is.EqualTo(expected))

[<Test>]
let ``Cross`` () = 
    let actual = 
        ["  *  ";
         "  *  ";
         "*****";
         "  *  ";
         "  *  "]
        |> concat

    let expected = 
        [" 2*2 ";
         "25*52";
         "*****";
         "25*52";
         " 2*2 "]
        |> concat

    Assert.That(annotate actual, Is.EqualTo(expected))

File: scala/minesweeper/src/test/scala/MinesweeperTest.scala
import org.scalatest.{Matchers, FunSuite}

/** @version 1.1.0 */
class MinesweeperTest extends FunSuite with Matchers {

  test("no rows") {
    Minesweeper.annotate(List()) should be(List())
  }

  test("no columns") {
    Minesweeper.annotate(List("")) should be(List(""))
  }

  test("no mines") {
    Minesweeper.annotate(List("   ",
                              "   ",
                              "   ")) should be(
      List("   ",
           "   ",
           "   "))
  }

  test("minefield with only mines") {
    Minesweeper.annotate(List("***",
                              "***",
                              "***")) should be(
      List("***",
           "***",
           "***"))
  }

  test("mine surrounded by spaces") {
    Minesweeper.annotate(List("   ",
                              " * ",
                              "   ")) should be(
      List("111",
           "1*1",
           "111"))
  }

  test("space surrounded by mines") {
    Minesweeper.annotate(List("***",
                              "* *",
                              "***")) should be(
      List("***",
           "*8*",
           "***"))
  }

  test("horizontal line") {
    Minesweeper.annotate(List(" * * ")) should be(List("1*2*1"))
  }

  test("horizontal line, mines at edges") {
    Minesweeper.annotate(List("*   *")) should be(List("*1 1*"))
  }

  test("vertical line") {
    Minesweeper.annotate(List(" ",
                              "*",
                              " ",
                              "*",
                              " ")) should be(
      List("1",
           "*",
           "2",
           "*",
           "1"))
  }

  test("vertical line, mines at edges") {
    Minesweeper.annotate(List("*",
                              " ",
                              " ",
                              " ",
                              "*")) should be(
      List("*",
           "1",
           " ",
           "1",
           "*"))
  }

  test("cross") {
    Minesweeper.annotate(List("  *  ",
                              "  *  ",
                              "*****",
                              "  *  ",
                              "  *  ")) should be(
      List(" 2*2 ",
           "25*52",
           "*****",
           "25*52",
           " 2*2 "))
  }

  test("large minefield") {
    Minesweeper.annotate(List(" *  * ",
                              "  *   ",
                              "    * ",
                              "   * *",
                              " *  * ",
                              "      ")) should be(
      List("1*22*1",
           "12*322",
           " 123*2",
           "112*4*",
           "1*22*2",
           "111111"))
  }
}
